### PR TITLE
Correctly persists Release upgrade failure

### PR DESCRIPTION
**Pull request by redbaron**
When release upgrade fails, updatedRelease is already created
in a storage by *ReleaseServer.UpdateRelease, therefore we should
be updating it's status, not creating it again.

### CODE REVIEW

**Comment by redbaron** [comment]
Here is how it looks like without a fix:


```
[tiller] 2017/09/22 11:00:02 creating updated release for prometheus-operator
FIRST >>>>>> [tiller:storage] 2017/09/22 11:00:02 creating release "prometheus-operator.v2"
[tiller] 2017/09/22 11:00:03 performing update for prometheus-operator
[tiller] 2017/09/22 11:00:03 executing 1 pre-upgrade hooks for prometheus-operator
[tiller] 2017/09/22 11:00:03 hooks complete for pre-upgrade prometheus-operator
[tiller:kubeClient] 2017/09/22 11:00:03 building resources from updated manifest
[tiller:kubeClient] 2017/09/22 11:00:03 checking 4 resources for changes
[tiller:kubeClient] 2017/09/22 11:00:03 Looks like there are no changes for ServiceAccount "prometheus-operator-prometheus-operator"
[tiller:kubeClient] 2017/09/22 11:00:04 Looks like there are no changes for ClusterRoleBinding "prometheus-operator-prometheus-operator"
[tiller] 2017/09/22 11:00:04 warning: Upgrade "prometheus-operator" failed: no resource with the name "prometheus-operator-prometheus-operator" found
[tiller:storage] 2017/09/22 11:00:04 updating release "prometheus-operator.v1"
SECOND >>>> [tiller:storage] 2017/09/22 11:00:05 creating release "prometheus-operator.v2"
[tiller] 2017/09/22 11:00:05 warning: Failed to record release prometheus-operator: release: "prometheus-operator.v2" already exists
```

**Comment by bacongobbler** [comment]
cc'ing @fibonacci1729 to review this since this package was his baby

**Comment by bacongobbler** [comment]
Any way you can write a unit test under release_update_test.go to verify this fix as well? Thanks!

**Comment by redbaron** [comment]
I'll add tests shortly

**Comment by redbaron** [comment]
Tests were in place, but Memory storage driver was storing pointers passed to it as arguments, so updates to them were implicitly updating stored Release. I have changed Memory storage driver to store deep copy,  so more store operation errors are visible

**Comment by redbaron** [comment]
@bacongobbler all tests are green now and actually test the thing, can you have a look? Change is fairly straightforward

**Comment by thomastaylor312** [comment]
@redbaron If @fibonacci1729 can't get around to this, I'll try to get to it on Friday or Monday (and if we decide to do a bug fix release, we can pull this in). Thank you for catching all of these bugs as of late

File: pkg/tiller/release_update_test.go [hunk 1]
```diff
@@ -285,3 +285,16 @@ func TestUpdateReleaseNoChanges(t *testing.T) {
 		t.Fatalf("Failed updated: %s", err)
 	}
 }
+
+func compareStoredAndReturnedRelease(t *testing.T, rs ReleaseServer, res services.UpdateReleaseResponse) *release.Release {
+	storedRelease, err := rs.env.Releases.Get(res.Release.Name, res.Release.Version)
```
Comment (bacongobbler) [inline pkg/tiller/release_update_test.go]: might want to add `t.Helper()` here so we get the correct line number when the test fails. This is a recent new feature in Go 1.9.

https://golang.org/pkg/testing/#B.Helper

Comment (redbaron) [reply pkg/tiller/release_update_test.go]: Once minimum Go version requirements are moved to 1.9 , it can be added.

Comment (bacongobbler) [reply pkg/tiller/release_update_test.go]: #2990 does that in doc form, but this isn't a blocker for this PR.

**Review by fibonacci1729 — APPROVED** [review]

**Pull request merged**